File: backend/app/schemas/well.py
from pydantic import BaseModel, Field
from datetime import date, datetime
from uuid import UUID


class WellCreate(BaseModel):
    well_name: str = Field(..., min_length=1, max_length=255)
    api_number: str | None = None
    uwi: str | None = None
    project_id: UUID | None = None

    # Location
    latitude: float | None = None
    longitude: float | None = None
    county: str | None = None
    state_province: str | None = None
    country: str = "US"
    basin: str | None = None
    field_name: str | None = None

    # Characteristics
    well_type: str = "oil"
    well_status: str = "active"
    orientation: str = "vertical"
    formation: str | None = None
    operator: str | None = None

    # Dates
    spud_date: date | None = None
    completion_date: date | None = None
    first_prod_date: date | None = None

    # Parameters
    total_depth: float | None = None
    lateral_length: float | None = None
    perf_top: float | None = None
    perf_bottom: float | None = None
    num_stages: int | None = None

    # Reservoir
    initial_pressure: float | None = None
    reservoir_temp: float | None = None
    porosity: float | None = None
    water_saturation: float | None = None
    net_pay: float | None = None
    permeability: float | None = None

    notes: str | None = None
    tags: list[str] | None = None


class WellUpdate(BaseModel):
    well_name: str | None = None
    api_number: str | None = None
    uwi: str | None = None
    project_id: UUID | None = None
    latitude: float | None = None
    longitude: float | None = None
    county: str | None = None
    state_province: str | None = None
    country: str | None = None
    basin: str | None = None
    field_name: str | None = None
    well_type: str | None = None
    well_status: str | None = None
    orientation: str | None = None
    formation: str | None = None
    operator: str | None = None
    spud_date: date | None = None
    completion_date: date | None = None
    first_prod_date: date | None = None
    total_depth: float | None = None
    lateral_length: float | None = None
    perf_top: float | None = None
    perf_bottom: float | None = None
    num_stages: int | None = None
    initial_pressure: float | None = None
    reservoir_temp: float | None = None
    porosity: float | None = None
    water_saturation: float | None = None
    net_pay: float | None = None
    permeability: float | None = None
    notes: str | None = None
    tags: list[str] | None = None


class WellResponse(BaseModel):
    id: UUID
    team_id: UUID
    project_id: UUID | None
    well_name: str
    api_number: str | None
    uwi: str | None
    latitude: float | None
    longitude: float | None
    county: str | None
    state_province: str | None
    country: str
    basin: str | None
    field_name: str | None
    well_type: str
    well_status: str
    orientation: str
    formation: str | None
    operator: str | None
    spud_date: date | None
    completion_date: date | None
    first_prod_date: date | None
    total_depth: float | None
    lateral_length: float | None
    perf_top: float | None
    perf_bottom: float | None
    num_stages: int | None
    initial_pressure: float | None
    reservoir_temp: float | None
    porosity: float | None
    water_saturation: float | None
    net_pay: float | None
    permeability: float | None
    notes: str | None
    tags: list[str] | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/well.ts
export type WellType = "oil" | "gas" | "oil_gas" | "injection";
export type WellStatus = "active" | "shut_in" | "plugged" | "drilling" | "completing";
export type WellOrientation = "vertical" | "horizontal" | "deviated";

export interface Well {
  id: string;
  team_id: string;
  project_id: string | null;
  well_name: string;
  api_number: string | null;
  uwi: string | null;

  // Location
  latitude: number | null;
  longitude: number | null;
  county: string | null;
  state_province: string | null;
  country: string;
  basin: string | null;
  field_name: string | null;

  // Characteristics
  well_type: WellType;
  well_status: WellStatus;
  orientation: WellOrientation;
  formation: string | null;
  operator: string | null;

  // Dates
  spud_date: string | null;
  completion_date: string | null;
  first_prod_date: string | null;

  // Parameters
  total_depth: number | null;
  lateral_length: number | null;
  perf_top: number | null;
  perf_bottom: number | null;
  num_stages: number | null;

  // Reservoir
  initial_pressure: number | null;
  reservoir_temp: number | null;
  porosity: number | null;
  water_saturation: number | null;
  net_pay: number | null;
  permeability: number | null;

  notes: string | null;
  tags: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface WellCreate {
  well_name: string;
  api_number?: string;
  uwi?: string;
  project_id?: string;
  latitude?: number;
  longitude?: number;
  county?: string;
  state_province?: string;
  country?: string;
  basin?: string;
  field_name?: string;
  well_type?: WellType;
  well_status?: WellStatus;
  orientation?: WellOrientation;
  formation?: string;
  operator?: string;
  spud_date?: string;
  completion_date?: string;
  first_prod_date?: string;
  total_depth?: number;
  lateral_length?: number;
  perf_top?: number;
  perf_bottom?: number;
  num_stages?: number;
  initial_pressure?: number;
  reservoir_temp?: number;
  porosity?: number;
  water_saturation?: number;
  net_pay?: number;
  permeability?: number;
  notes?: string;
  tags?: string[];
}

export type WellUpdate = Partial<WellCreate>;

export interface WellFilters {
  page?: number;
  per_page?: number;
  project_id?: string;
  well_status?: WellStatus;
  well_type?: WellType;
  basin?: string;
  search?: string;
}
